SQL/proc_descriptions/testing_extension.sql

```diff
@@ -0,0 +1,264 @@
+        FOR result_record IN
+            SELECT * FROM _runner(
+                findfuncs_recursive( $1, startup_pattern ),
+                findfuncs_recursive( $1, shutdown_pattern ),
+                findfuncs_recursive( $1, setup_pattern ),
+                findfuncs_recursive( $1, teardown_pattern ),
+                findfuncs_recursive( $1, $2, exclude_pattern )
+            )
+        LOOP
+            RETURN NEXT result_record;
+        END LOOP;
+
+        -- Raise the exception
+        RAISE EXCEPTION '__TAP_ROLLBACK__' USING ERRCODE = '22012';
+    EXCEPTION
+        WHEN SQLSTATE '22012' THEN
+            -- Catch the exception to prevent it from propagating
+            NULL; -- Do nothing
+    END;
+END;
+$$;
+
+-- Function: mob_group_runtests(text)
+-- This function is used to run a group of tests.
+-- Input parameters: text - the pattern to match the names of the functions to run.
+-- Flow:
+-- 1. Run alphabetically ordered startup functions once, which match the pattern "startup" + $1.
+-- 2. Enter a loop:
+--  2.1. Run alphabetically ordered setup functions once, which match the pattern "setup" + $1.
+--  2.2. Run alphabetically ordered test functions once, which match the pattern $1.
+--  2.3. Run alphabetically ordered teardown functions once, which match the pattern "teardown" + $1.
+-- 3. Run alphabetically ordered shutdown functions once, which match the pattern "shutdown" + $1.
+-- 4. Return the results of the tests.
+create or replace function mob_group_runtests(text) returns SETOF text
+    language plpgsql
+as
+$$
+DECLARE
+    startup TEXT := 'startup' || $1;
+    shutdown TEXT := 'shutdown' || $1;
+    setup TEXT := 'setup' || $1;
+    teardown TEXT := 'teardown' || $1;
+    test TEXT := 'test' || $1;
+    startup_pattern TEXT := '^' || startup;
+    shutdown_pattern TEXT := '^' || shutdown;
+    setup_pattern TEXT := '^' || setup;
+    teardown_pattern TEXT := '^' || teardown;
+    test_pattern TEXT := '^' || test;
+    exclude_pattern TEXT := '^(' || startup || '|' || shutdown || '|' || setup || '|' || teardown || ')';
+    result_record RECORD;
+BEGIN
+    -- A little note about raising exception in this function, it seems that `_runner` function
+    -- does not raise an exception, instead catching it and returning it as a record, so under
+    -- this assumption we raise "division by zero" AKA "22012" exception to rollback the transaction.
+    BEGIN -- begin transaction for later rollback
+    FOR result_record IN
+        SELECT * FROM _runner(
+            findfuncs_recursive( startup_pattern ),
+            findfuncs_recursive( shutdown_pattern ),
+            findfuncs_recursive( setup_pattern ),
+            findfuncs_recursive( teardown_pattern ),
+            findfuncs_recursive( test_pattern, exclude_pattern )
+        )
+    LOOP
+        RETURN NEXT result_record;
+    END LOOP;
+        -- Raise the exception
+        RAISE EXCEPTION '__TAP_ROLLBACK__' USING ERRCODE = '22012';
```

Comment:
Yes, I've run basic pgTap function `runtests()` multiple times and the one thing I've noticed, that every change made by `^startup` functions to database are not reversed (haven't tested it with `^shutdown` functions, but I think it doesn't revert modifications too), that's why I've come up with this algorithm - raise an exception to catch it immediately, this way postgres ROLLBACK function transaction, which forces any changes made by `^startup` functions to revert

Reply:
I see, so this is the only way, can't we rollback manually?

Reply:
[Postgres does not support such procedures as `ROLLBACK`, `COMMIT` in functions](https://stackoverflow.com/questions/48277519/how-to-use-commit-and-rollback-in-a-postgresql-function)
```
You cannot use transaction statements like SAVEPOINT, COMMIT or ROLLBACK in a function. [The documentation](https://www.postgresql.org/docs/current/plpgsql-transactions.html) says:

In procedures invoked by the CALL command as well as in anonymous code blocks (DO command), it is possible to end transactions using the commands COMMIT and ROLLBACK.
```

Reply:
if it works within a procedure, why not implement the test executor as a procedure?

Reply:
`mob_group_runtests()` function is supposed to return SETOF TEXT, which is basically a set of descriptions of how a particular subtest finished. Not really sure how we should show results of tests if not by returning texts.

Reply:
well I'm not saying that we should not return values, but why not return them from a procedure, if we get an additional advantage of the rollback support?

Reply:
Procedure does not provide returning values feature, does not it? I though it was the key concept for a `procedure`

Reply:
you can return from a procedure using an INOUT parameter (see [argmode](https://www.postgresql.org/docs/current/sql-createprocedure.html))

Reply:
or just OUT...

Reply:
It seems like its taking some time to restructure `mob_group_runtest` into _procedure_ (some technical details, like procedure not supporting SETOF TEXT argument and so on). Including that we already have working system, im not sure, that we should invest time into this. One thing I could do is to replace errcode from current 'division by zero' to some custom one, which is not included in postgres documentation, so it would not overlap.